### ISSUE TITLE
Unknown layer type error when loading any layer

### ISSUE THREAD

**Issue by rrshaban** [ethereon/caffe-tensorflow#5]
I've been trying to load [the standard VGG16 models and prototxt](https://gist.github.com/ksimonyan/211839e770f7b538e2d8#file-readme-md), but am getting the following error when trying to load the very first layer: 

`Unknown layer type encountered: None`

The error is thrown in `core.py:GraphBuilder:make_node`, which has the following lines:
```
   kind = NodeKind.map_raw_kind(layer.type)
   if kind is None:
      raise KaffeError('Unknown layer type encountered: %s'%kind)
```
`layer.type` here is an integer (`4` – no idea where that's coming from), but `NodeKind.map_raw_kind` is trying to match the layer name (e.g. "Convolution" to LAYER_TYPES, which contains a list of Caffe layer types). Naturally, `4` isn't in LAYER_TYPES, so it throws the error. 

I can't figure out where this 4, instead of "Convolution", is coming from – tracing that back led me to kaffe/caffepb, which I don't understand. The layer object looks like this: `<kaffe.caffepb.V1LayerParameter object at 0x1064c2a28>`, and this is what `print(layer)` outputs right before the call to `layer.type` returns 4:
```
bottom: "data"
top: "conv1_1"
name: "conv1_1"
type: CONVOLUTION
convolution_param {
  num_output: 64
  pad: 1
  kernel_size: 3
}
```

Any thoughts as to what's going on? I imagine the wrong thing is being assigned to type, but I can't figure out where. 

As an aside, the names in `LAYER_DESCRIPTORS` are all in CamelCase, while the ones in the official prototxts are in all-caps – I've modified manually as a quick fix, but might submit a .lower() fix if I have time today.

**Comment by ethereon** [comment]
It sounds like you're using the old Caffe format (the original VGG16 gist uses the old format as well).
Try upgrading the prototxt and caffemodel to the most current version, as described in the readme.

**Comment by arnavagharwal** [comment]
I'm also facing the same problem. I attempted a version upgrade using the scripts provided in the distribution, but getting the message that both the model, as well as the prototxt files are already in V1 format.
> E1125 18:37:36.422992 11962 upgrade_net_proto_binary.cpp:37] File already in V1 proto format: VGG_ILSVRC_16_layers.caffemodel

To be safe, I also ran a checksum on both input file to, and the output files from the `upgrade_net_proto_binary.bin` Caffe script, and they are identical.

> Try upgrading the prototxt and caffemodel to the most current version, as described in the readme.

Is that higher than V1?

**Comment by ethereon** [comment]
@arnavagharwal: The V1 format should be fine (here's a quick sanity check: your prototxt file should consist of the singular `layer`, rather than `layers`).

Could you describe the exact error that you encounter?

**Comment by arnavagharwal** [comment]
@ethereon: Thank you for responding.
As described by @rrshaban, I get the error
>Error encountered: Unknown layer type encountered: None
Tracing the source leads to the same conclusion as his.

My prototxt file contains the plural `layers`. Based on what you say, this is clearly the outdated version, so not sure why `upgrade_net_proto_binary.bin` is refusing to upgrade. I am using an older version of Caffe (don't know version number), but I will repeat the upgrade attempt using the latest release. Will update this thread with the outcome, once that is done.

**Comment by arnavagharwal** [comment]
With the latest Caffe commit BVLC/caffe@8e8d97d6206cac99eae3c16baaa2275a14e64ca7, both the model, as well as the prototxt files were upgraded. Before that, I also tried with version 0.99 on another machine, and no upgrades were performed. Not sure why this is the case. Perhaps, I should file a bug report?

Great job with the conversion! It's a Godsend for early Tensorflow adopters.

Lastly, if you are interested in even more public service, you could consider adding an example feature extraction script, as done in a similar repo ry/tensorflow-vgg16@4c454cf5ed89a0a57c29ed71fa81497672644d96. Thank you for your time!

**Comment by ethereon** [comment]
@arnavagharwal 0.99 is over a year old, so I wouldn't be surprised if things have changed since then.

@rrshaban I'm closing this issue since it seems solved. If the problem persists, feel free to create another issue.
